File: src/scraper/rate_limiter.py
"""
Keiba AI - Rate Limiter

Async token bucket rate limiter to avoid IP blocks when scraping.
"""

import asyncio
import random
import time
from dataclasses import dataclass, field
from typing import Optional

from .config import ScraperConfig


@dataclass
class RateLimiter:
    """
    Async token bucket rate limiter with random delays.

    Implements a token bucket algorithm where tokens are replenished
    at a fixed rate. Also adds random delays between requests.

    Usage:
        limiter = RateLimiter.from_config(config)
        async with limiter:
            # Make request here
            pass
    """

    requests_per_minute: int = 20
    min_delay: float = 1.5
    max_delay: float = 3.0
    _tokens: float = field(default=0.0, init=False)
    _last_update: float = field(default_factory=time.monotonic, init=False)
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock, init=False)

    def __post_init__(self):
        # Start with full bucket
        self._tokens = float(self.requests_per_minute)
        self._last_update = time.monotonic()

    @classmethod
    def from_config(cls, config: Optional[ScraperConfig] = None) -> "RateLimiter":
        """Create rate limiter from scraper config."""
        config = config or ScraperConfig()
        return cls(
            requests_per_minute=config.requests_per_minute,
            min_delay=config.min_delay_seconds,
            max_delay=config.max_delay_seconds,
        )

    @property
    def _replenish_rate(self) -> float:
        """Tokens per second."""
        return self.requests_per_minute / 60.0

    def _replenish(self) -> None:
        """Replenish tokens based on elapsed time."""
        now = time.monotonic()
        elapsed = now - self._last_update
        self._tokens = min(
            float(self.requests_per_minute),
            self._tokens + elapsed * self._replenish_rate,
        )
        self._last_update = now

    async def acquire(self) -> None:
        """
        Acquire a token, waiting if necessary.

        This method is thread-safe and will wait if no tokens are available.
        """
        async with self._lock:
            self._replenish()

            # Wait if no tokens available
            if self._tokens < 1.0:
                wait_time = (1.0 - self._tokens) / self._replenish_rate
                await asyncio.sleep(wait_time)
                self._replenish()

            # Consume token
            self._tokens -= 1.0

            # Add random delay
            delay = random.uniform(self.min_delay, self.max_delay)
            await asyncio.sleep(delay)

    async def __aenter__(self) -> "RateLimiter":
        """Acquire token on context enter."""
        await self.acquire()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Nothing to do on exit."""
        pass

    @property
    def available_tokens(self) -> float:
        """Get current number of available tokens (for debugging)."""
        self._replenish()
        return self._tokens


async def test_rate_limiter():
    """Test rate limiter functionality."""
    limiter = RateLimiter(requests_per_minute=10, min_delay=0.1, max_delay=0.2)

    print(f"Starting with {limiter.available_tokens:.1f} tokens")

    for i in range(5):
        start = time.monotonic()
        async with limiter:
            elapsed = time.monotonic() - start
            print(f"Request {i + 1}: waited {elapsed:.2f}s, tokens: {limiter.available_tokens:.1f}")

    print("Rate limiter test successful!")


if __name__ == "__main__":
    asyncio.run(test_rate_limiter())

File: src/api/src/scraper/rate_limiter.rs
//! Rate limiter using token bucket algorithm.

use std::sync::Arc;
use tokio::sync::Mutex;
use tokio::time::{Duration, Instant};

/// Token bucket rate limiter
pub struct RateLimiter {
    state: Arc<Mutex<RateLimiterState>>,
}

struct RateLimiterState {
    tokens: f64,
    last_update: Instant,
    max_tokens: f64,
    refill_rate: f64, // tokens per second
    min_delay: Duration,
    max_delay: Duration,
}

impl RateLimiter {
    /// Create a new rate limiter
    ///
    /// # Arguments
    /// * `requests_per_minute` - Maximum requests per minute
    /// * `min_delay_secs` - Minimum delay between requests
    /// * `max_delay_secs` - Maximum delay between requests
    pub fn new(requests_per_minute: u32, min_delay_secs: f64, max_delay_secs: f64) -> Self {
        let max_tokens = requests_per_minute as f64;
        let refill_rate = requests_per_minute as f64 / 60.0;

        Self {
            state: Arc::new(Mutex::new(RateLimiterState {
                tokens: max_tokens,
                last_update: Instant::now(),
                max_tokens,
                refill_rate,
                min_delay: Duration::from_secs_f64(min_delay_secs),
                max_delay: Duration::from_secs_f64(max_delay_secs),
            })),
        }
    }

    /// Create with default settings (60 req/min, 0.5-1.0s delay)
    pub fn default_limiter() -> Self {
        Self::new(60, 0.5, 1.0)
    }

    /// Acquire a token, waiting if necessary
    pub async fn acquire(&self) {
        let delay = {
            let mut state = self.state.lock().await;

            // Refill tokens
            let now = Instant::now();
            let elapsed = now.duration_since(state.last_update).as_secs_f64();
            state.tokens = (state.tokens + elapsed * state.refill_rate).min(state.max_tokens);
            state.last_update = now;

            // Check if we have tokens
            if state.tokens >= 1.0 {
                state.tokens -= 1.0;
                // Return random delay between min and max
                let delay_range = state.max_delay - state.min_delay;
                let random_factor = rand_delay();
                state.min_delay + delay_range.mul_f64(random_factor)
            } else {
                // Wait for token to become available
                let wait_time = (1.0 - state.tokens) / state.refill_rate;
                state.tokens = 0.0;
                Duration::from_secs_f64(wait_time) + state.min_delay
            }
        };

        tokio::time::sleep(delay).await;
    }
}

/// Generate a pseudo-random delay factor (0.0 - 1.0)
fn rand_delay() -> f64 {
    use std::time::SystemTime;
    let nanos = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .subsec_nanos();
    (nanos % 1000) as f64 / 1000.0
}
